Add Cross-Encoder Reranking to Search Agent

pranath-reddy opened the pull request:
### Summary :memo:
This PR integrates reranking capabilities into the search agent architecture by adding cross-encoder reranking support at the base SearchAgent level. The reranker improves search result quality by reordering results based on semantic relevance to the query, and can be optionally disabled by setting the reranker type to "none".

### Details
- Added reranker configuration fields to `SearchAgentConfig`:
  - `reranker_type: RerankerType` with default value "none" to specify the type of reranker
  - `reranker_config: RerankerToolConfig` with default factory for cross-encoder configuration (model: "cross-encoder/ms-marco-MiniLM-L12-v2")
- Integrated reranker imports and tooling from `akd.tools.reranker` module:
  - `RerankerTool` for executing reranking operations
  - `RerankerToolConfig` for configuration
  - `RerankerType` enum for specifying reranker model
  - `create_reranker` factory function
  - `RerankerToolInputSchema` for input validation
- Added reranker initialization in `SearchAgent.__init__()` method, making it available to all search agent subclasses
- Updated `CodeSearchAgentConfig` to inherit reranker configuration with default type "cross_encoder"
- Implemented reranking logic in `ControlledSearchAgent._run()` method:
  - Reranks accumulated results after each search iteration
  - Uses `RerankerToolInputSchema` to structure reranker input with query and results
  - Replaces `all_results` with reranked results to improve result ordering for subsequent iterations

### Checks
- [x] Tested Changes
- [ ] Stakeholder Approval

pranath-reddy committed:
Add reranker to controlled agent

- Added reranker to the controlled agent flow
- instantiate the reranker inside the class constructor
- if reranking is enabled, rerank the `all_results` list based on the original input query

pranath-reddy committed:
set flag

- enable reranking in code search agent

pranath-reddy committed:
Add val scripts

- Added validation scripts

pranath-reddy committed:
update agent val script

- updated configs of the tools

pranath-reddy committed:
Remove eval scripts

- move eval scripts to `code-search-eval` branch

pranath-reddy committed:
Merge remote-tracking branch 'origin/develop' into feature/code-agent-rerank

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -66,6 +66,12 @@ class CodeSearchAgentConfig(ControlledSearchAgentConfig):
     relevancy_prompt: str = Field(default=CODE_RELEVANCY_PROMPT, description="System prompt for relevancy agent")
 
+    # Reranker configuration
+    search_result_reranking: bool = Field(
```
Maybe just doing reranking parameter? I think we don't even need this because if we have `reranker_type` already, and setting it to `none` or `nope`, it iwll use NoOp reranker (no changes). 

in create_reranker function...

```python
    if reranker_type in ("identity", "no_op", "nope", "none"):
        return NoOpRerankerTool(config=config, debug=debug)
```

NISH1001 commented on akd/agents/search/controlled.py:
```diff
@@ -120,6 +127,22 @@ class ControlledSearchAgentConfig(LitSearchAgentConfig):
+    search_result_reranking: bool = Field(
+        default=False,
+        description="Enable reranking of search results.",
+    )
```
I don't think we need it. This can easily be done through `reranker_type` and setting to `none` or `nope` will do it.

create_reranker function
```python
    if reranker_type in ("identity", "no_op", "nope", "none"):
        return NoOpRerankerTool(config=config, debug=debug)

```

pranath-reddy replied:
Ah yes. got it. btw why do we have `nope` in there XD

NISH1001 replied:
It was an easter egg. lmao. i like to put some hidden funny things in code bases.

NISH1001 commented on akd/agents/search/controlled.py:
```diff
@@ -850,6 +879,14 @@ async def _arun(
 
             all_results.extend(current_results)
 
+            if self.config.search_result_reranking:
```
We could avoid this conditional check and direclty apply reranking. Because if it was `none` or `nope` type or osmething, doing `reranker.arun(...)` won't affect anything with original list.
```

reranker_input = RerankerToolInputSchema(...)
reranked_results = await self.reranker.arun(reranker_input)
all_results = reranked_results.results
```

This avoids conditional and uses state-machine-like pattern based on reranker_type.

pranath-reddy committed:
address PR comments

- Remove redundant reranking flag since reranker already supports `none` (No reranking)

pranath-reddy committed:
update search base agent

- Initialize the reranker at base agent instead of controlled so that the reranker is reusable across all search agents.
